apiserversdk/proxy.go

```diff
@@ -46,3 +58,30 @@ func WithFieldSelector(handler http.Handler, selectors ...string) http.Handler {
 		handler.ServeHTTP(w, r)
 	})
 }
+
+// requireKuberayService verifies that the requested service has the label "app.kubernetes.io/name=kuberay".
+// If the service is not found or does not have the correct label, it returns a 404 Not Found error.
+func requireKuberayService(handler http.Handler, k8sClient *kubernetes.Clientset) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		namespace, serviceSchemeNamePort := r.PathValue("namespace"), r.PathValue("service")
+		_, serviceName, _, valid := net.SplitSchemeNamePort(serviceSchemeNamePort)
+		if !valid {
+			http.Error(w, "invalid service format: "+serviceSchemeNamePort, http.StatusBadRequest)
+			return
+		}
+
+		services, err := k8sClient.CoreV1().Services(namespace).List(r.Context(), metav1.ListOptions{
```

Comment:
Not sure why do we use list operation here, it's usually backed by a range query with two predicates to etcd.
- Range query is generally slow than point query
- What if there're a bunch of items returned?

At we could apply limit in the list option?

Reply:
Just to simplify our checks by leveraging label selectors. Kubernetes apiserver has an optimization for `metadata.name=` case:
https://github.com/kubernetes/apiserver/blob/9e1714f2001f6b644d8e5fa86534b10044b60f98/pkg/registry/generic/registry/store.go#L414-L421

It won't return multiple items.

Reply:
Good to know, thanks for sharing!